tests/conftest.py

```
@@ -55,6 +56,24 @@ def parse_paragraph(xml: str):
 
+def count_dom_walks(monkeypatch) -> list[str]:
+    """Record the tag of every full-document getElementsByTagName call.
+
+    Returns a list that grows as walks happen, so a test can pin that the
+    number of full-DOM walks stays constant in the operation count. Only
+    Document-level walks are counted; paragraph-local ones run on Elements.
+    """
+    walks: list[str] = []
+    original = minidom.Document.getElementsByTagName
+
+    def counting(self, name):
```

Ruff (0.15.21)

[warning] Missing return type annotation for private function `counting`

(ANN202)